src/epics.c

```diff
@@ -0,0 +1,575 @@
+
+  switch (args.op) {
+  case CA_OP_CONN_UP:
+    handle_conn_up(p);
+    break;
+  case CA_OP_CONN_DOWN:
+    handle_conn_down(p);
+    break;
+  }
+}
+
+static void *epics_thread(void *args) {
+  long ret = ca_context_create(ca_disable_preemptive_callback);
+  if (ret != ECA_NORMAL) {
+    // FIXME: report error back to start_thread()
+    ERROR("epics plugin: CA error %s occurred while trying to start channel "
+          "access",
+          ca_message(ret));
+    return (void *)1;
+  }
+
+  ca_replace_printf_handler(&printf_handler);
+
+  for (size_t i = 0; i < epics_plugin.pvs_num; ++i) {
+    struct pv *p = &epics_plugin.pvs[i];
+
+    ret = ca_create_channel(p->name, &connection_handler, p, 0, &p->id);
+    if (ret != ECA_NORMAL) {
+      ERROR("epics plugin: CA error %s occurred while trying to create channel "
+            "\"%s\"",
+            ca_message(ret), p->name);
+      ret = 1;
+      goto error;
+    }
+  }
+
+  const double timeout = 2.0;
+  while (epics_plugin.thread_loop != 0) {
+    ca_pend_event(timeout);
+  }
+
+error:
+  for (size_t i = 0; i < epics_plugin.pvs_num; ++i) {
+    struct pv *p = &epics_plugin.pvs[i];
+
+    if (p->eid) {
+      ca_clear_subscription(p->eid);
+      p->eid = 0;
+    }
+    ca_clear_channel(p->id);
+  }
+
+  ca_context_destroy();
+
+  return (void *)ret;
+}
+
+static int start_thread(void) {
+  pthread_mutex_lock(&epics_plugin.lock);
+
+  int ret = 0;
+  if (epics_plugin.thread_loop != 0) {
+    goto epics_unlock;
+  }
+
+  epics_plugin.thread_loop = 1;
+  ret = plugin_thread_create(&epics_plugin.thread_id, epics_thread, (void *)0,
+                             "epics");
+  if (ret != 0) {
+    epics_plugin.thread_loop = 0;
+    ERROR("epics plugin: Starting thread failed: %d", ret);
+
+    goto epics_unlock;
+  }
+
+  // FIXME: wait untill ca_context_create success
+
+epics_unlock:
+  pthread_mutex_unlock(&epics_plugin.lock);
+
+  return ret;
+}
+
+static int stop_thread(void) {
+  pthread_mutex_lock(&epics_plugin.lock);
+
+  epics_plugin.thread_loop = 0;
+
+  pthread_mutex_unlock(&epics_plugin.lock);
+
+  return pthread_join(epics_plugin.thread_id, NULL);
+}
+
+static int epics_config_metric(oconfig_item_t *ci, struct pv *p,
+                               size_t family_index) {
+  if (cf_util_get_string(ci, &p->name) != 0 || p->name == NULL) {
+    ERROR("epics plugin: Wrong metric configuration");
+
+    return -1;
+  }
+
+  p->family_index = family_index;
+
+  return 0;
+}
+
+static int epics_config_label(oconfig_item_t *ci, struct pv *p,
+                              size_t family_index) {
+  if (cf_util_get_string(ci, &p->name) != 0 || p->name == NULL) {
+    ERROR("epics plugin: Wrong label configuration");
+
+    return -1;
+  }
+
+  p->family_index = family_index;
+  p->is_label = 1;
+
+  return 0;
+}
+
+static int epics_config_metric_family(oconfig_item_t *ci, size_t family_index) {
+  metric_family_t *fam = &epics_plugin.families[family_index];
```

Comment:
Any particular reason for passing and (storing to `pv` struct) an index to a global array instead of  a `metric_family_t` pointer?

Reply:
Yes, `families` are reallocated at each config function invocation. Storing the pointer would easily make it dangling.